cli/src/git_util.rs

```diff
@@ -138,32 +138,117 @@ fn get_ssh_keys(_username: &str) -> Vec<PathBuf> {
     paths
 }
 
+// Based on Git's implementation: https://github.com/git/git/blob/43072b4ca132437f21975ac6acc6b72dc22fd398/sideband.c#L178
+pub struct GitSidebandProgressMessageWriter {
+    display_prefix: &'static [u8],
+    suffix: &'static [u8],
+    scratch: Vec<u8>,
+}
+
+impl Default for GitSidebandProgressMessageWriter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GitSidebandProgressMessageWriter {
+    pub fn new() -> Self {
+        GitSidebandProgressMessageWriter {
+            display_prefix: "remote: ".as_bytes(),
+            suffix: if std::io::stderr().is_terminal() {
+                "\x1B[K"
+            } else {
+                "        "
+            }
+            .as_bytes(),
+            scratch: Vec::new(),
+        }
+    }
+
+    pub fn write(&mut self, ui: &Ui, progress_message: &[u8]) {
+        let mut index = 0;
+        // Append a suffix to each nonempty line to clear the end of the screen line.
+        loop {
+            let Some(i) = progress_message[index..]
+                .find_char('\n')
+                .or_else(|| progress_message[index..].find_char('\r'))
+                .map(|i| index + i)
+            else {
```

Comment:
nit: `.iter().position(|&c| c == b'\r' || c == b'\n')`?

Maybe you can also use `.split_inclusive(..)`. (untested)
```rust
let mut lines = .split_inclusive(..);
let last_line = lines.next_back();
for line in lines {
    ...
```
https://doc.rust-lang.org/std/primitive.slice.html#method.split_inclusive

Reply:
I switched to `.iter().position()`; opted against `split_inclusive` since we would need to check if the `last_line` ended with `\r` or `\n` as well.